src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -346,6 +367,82 @@ public void Load()
             return (endpointsToStop, endpointsToStart);
         }
 
+        private ValueTask<SslServerAuthenticationOptions> SniCallback(
+            HttpsConnectionAdapterOptions httpsOptions,
+            HttpProtocols endpointDefaultHttpProtocols,
+            EndpointConfig endpoint,
+            SslClientHelloInfo clientHelloInfo)
+        {
+            const string wildcardHost = "*";
+            const string wildcardPrefix = "*.";
+
+            var sslServerOptions = new SslServerAuthenticationOptions();
+            var serverName = clientHelloInfo.ServerName;
+            SniConfig sniConfig = null;
+
+            if (!string.IsNullOrEmpty(serverName))
+            {
+                foreach (var (name, sniConfigCandidate) in endpoint.SNI)
```

Comment:
Pre-build a `Dictionary<string, SslServerAuthenticationOptions>` ignore case collection when loading config. Consider setting a flag that says if the SNI contains any top level wildcards (`*.domain`) or a global wildcard (`*`). That way you can avoid enumerating the dictionary if there aren't any.

Matching:
1. If there's no serverName check the dictionary for a top level wildcard `*`. If not present then fail.
1. Direct lookup in the dictionary (fastest)
1. Scan the dictionary for the most specific wildcard `*.domain`, or top level wildcard `*`.
1. Fail if there's no match or wildcard in the dictionary.